### PR TITLE
tui: support a way to change sort key in TUI report mode

### PR DESCRIPTION
**Pull request by hjongkim**
Added function to sort by key when pressing 's' key in TUI report mode.

Fixed: #907

Signed-off-by: Hyoungjong Kim <exist91@gmail.com>

### CODE REVIEW

**Comment by honggyukim** [comment]
It seems working fine, but it'd better to have a way to see which KEY is used for sorting.

In addition, please add a description in tui man page. Thanks!

File: cmds/tui.c [hunk 1]
```diff
@@ -2463,6 +2467,16 @@ static void tui_main_loop(struct opts *opts, struct uftrace_data *handle)
 				full_redraw = true;
 			}
 			break;
+		case 's':
+			if (!tui_window_change(win, &report->win)) {
+				report_setup_sort(report_sort_key[curr_sort_key++]);
```
Comment (namhyung) [inline cmds/tui.c]: But this will ignore the first key press.

File: cmds/tui.c [hunk 2]
```diff
@@ -2463,6 +2467,16 @@ static void tui_main_loop(struct opts *opts, struct uftrace_data *handle)
...
+				curr_sort_key %= (sizeof(report_sort_key)/sizeof(report_sort_key[0]));
```
[the remark this answers is not in the document]
Comment (namhyung) [reply cmds/tui.c]: You can use `ARRAY_SIZE()` macro.

**Comment by honggyukim** [comment]
Please squash the commits into a single one. Thanks!

**Comment by honggyukim** [comment]
@hjongkim What about the following change? It's based on your current branch.
```diff
diff --git a/cmds/tui.c b/cmds/tui.c
index 5365e7e..74259fe 100644
--- a/cmds/tui.c
+++ b/cmds/tui.c
@@ -160,7 +160,8 @@ static const char *help[] = {
        "q             Quit",
 };

-const char *report_sort_key[] = { OPT_SORT_KEYS, "self", "call" };
+static const char *report_sort_key[] = { OPT_SORT_KEYS, "self", "call" };
+static int curr_sort_key = 0;

 static void init_colors(void)
 {
@@ -1332,8 +1333,10 @@ static bool win_search_report(struct tui_window *win, void *node, char *str)
 static void win_header_report(struct tui_window *win, struct uftrace_data *handle)
 {
        int w = 46;
+       char header[][32] = { " Total Time", " Self Time", " Calls" };

-       printw("  %10s  %10s  %10s  %s", "Total Time", "Self Time", "Calls", "Function");
+       header[curr_sort_key][0] = '*';
+       printw(" %11s %11s %11s  %s", header[0], header[1], header[2], "Function");
        if (COLS > w)
                printw("%*s", COLS - w, "");
 }
@@ -2323,7 +2326,6 @@ static inline void cancel_search()
 static void tui_main_loop(struct opts *opts, struct uftrace_data *handle)
 {
        int key = 0;
-       int curr_sort_key = 0;
        bool full_redraw = true;
        struct tui_graph *graph;
        struct tui_report *report;
@@ -2469,9 +2471,11 @@ static void tui_main_loop(struct opts *opts, struct uftrace_data *handle)
                        break;
                case 's':
                        if (!tui_window_change(win, &report->win)) {
-                               report_setup_sort(report_sort_key[curr_sort_key++]);
-                               curr_sort_key %= (sizeof(report_sort_key)/sizeof(report_sort_key[0]));
-                               report_sort_nodes(&tui_report.name_tree, &tui_report.sort_tree);
+                               curr_sort_key = (curr_sort_key + 1) %
+                                               ARRAY_SIZE(report_sort_key);
+                               report_setup_sort(report_sort_key[curr_sort_key]);
+                               report_sort_nodes(&tui_report.name_tree,
+                                                 &tui_report.sort_tree);

                                tui_window_move_home(win);
                                full_redraw = true;
```
You can download the diff file here - [tui-report-sort.diff](https://github.com/namhyung/uftrace/files/3757083/tui-report-sort.diff.txt)

**Comment by hjongkim** [comment]
> @hjongkim What about the following change? It's based on your current branch.
> 
> ```diff
> diff --git a/cmds/tui.c b/cmds/tui.c
> index 5365e7e..74259fe 100644
> --- a/cmds/tui.c
> +++ b/cmds/tui.c
> @@ -160,7 +160,8 @@ static const char *help[] = {
>         "q             Quit",
>  };
> 
> -const char *report_sort_key[] = { OPT_SORT_KEYS, "self", "call" };
> +static const char *report_sort_key[] = { OPT_SORT_KEYS, "self", "call" };
> +static int curr_sort_key = 0;
> 
>  static void init_colors(void)
>  {
> @@ -1332,8 +1333,10 @@ static bool win_search_report(struct tui_window *win, void *node, char *str)
>  static void win_header_report(struct tui_window *win, struct uftrace_data *handle)
>  {
>         int w = 46;
> +       char header[][32] = { " Total Time", " Self Time", " Calls" };
> 
> -       printw("  %10s  %10s  %10s  %s", "Total Time", "Self Time", "Calls", "Function");
> +       header[curr_sort_key][0] = '*';
> +       printw(" %11s %11s %11s  %s", header[0], header[1], header[2], "Function");
>         if (COLS > w)
>                 printw("%*s", COLS - w, "");
>  }
> @@ -2323,7 +2326,6 @@ static inline void cancel_search()
>  static void tui_main_loop(struct opts *opts, struct uftrace_data *handle)
>  {
>         int key = 0;
> -       int curr_sort_key = 0;
>         bool full_redraw = true;
>         struct tui_graph *graph;
>         struct tui_report *report;
> @@ -2469,9 +2471,11 @@ static void tui_main_loop(struct opts *opts, struct uftrace_data *handle)
>                         break;
>                 case 's':
>                         if (!tui_window_change(win, &report->win)) {
> -                               report_setup_sort(report_sort_key[curr_sort_key++]);
> -                               curr_sort_key %= (sizeof(report_sort_key)/sizeof(report_sort_key[0]));
> -                               report_sort_nodes(&tui_report.name_tree, &tui_report.sort_tree);
> +                               curr_sort_key = (curr_sort_key + 1) %
> +                                               ARRAY_SIZE(report_sort_key);
> +                               report_setup_sort(report_sort_key[curr_sort_key]);
> +                               report_sort_nodes(&tui_report.name_tree,
> +                                                 &tui_report.sort_tree);
> 
>                                 tui_window_move_home(win);
>                                 full_redraw = true;
> ```
> 
> You can download the diff file here - [tui-report-sort.diff](https://github.com/namhyung/uftrace/files/3757083/tui-report-sort.diff.txt)

**Comment by hjongkim** [comment]
Thank you for helping me. I fixed it. Please check.

File: cmds/tui.c [hunk 3]
```diff
@@ -142,6 +142,7 @@ static const char *help[] = {
 	"g             Show call graph for this function",
 	"R             Show uftrace report",
 	"r             Show uftrace report for this function",
+	"s             Show uftrace sorted report by KEYs",
```
Comment (honggyukim) [inline cmds/tui.c]: What about this?
```suggestion
	"s             Show uftrace report sorted by KEYs",
```

File: cmds/tui.c [hunk 4]
```diff
@@ -2463,6 +2469,18 @@ static void tui_main_loop(struct opts *opts, struct uftrace_data *handle)
 				full_redraw = true;
 			}
 			break;
+		case 's':
+			if (!tui_window_change(win, &report->win)) {
+				curr_sort_key = (curr_sort_key + 1) %
+					ARRAY_SIZE(report_sort_key);
```
Comment (honggyukim) [inline cmds/tui.c]: Please give one more tab for indentation like below.
```suggestion
						ARRAY_SIZE(report_sort_key);
```

File: cmds/tui.c [hunk 5]
```diff
@@ -2463,6 +2469,18 @@ static void tui_main_loop(struct opts *opts, struct uftrace_data *handle)
...
+				report_setup_sort(report_sort_key[curr_sort_key]);
+				report_sort_nodes(&tui_report.name_tree,
+						&tui_report.sort_tree);
```
Comment (honggyukim) [inline cmds/tui.c]: Please give two more whitespaces to align with the previous line.
```suggestion
						  &tui_report.sort_tree);
```

File: doc/uftrace-tui.md [hunk 6]
```diff
@@ -147,6 +147,7 @@ Following keys can be used in the TUI window:
  * `g`:                   Show backtrace and call graph of the current function
  * `R`:                   Show uftrace report
  * `r`:                   Show uftrace report of the current function
+ * `s`:                   Show uftrace sorted report by KEYs
```
Comment (honggyukim) [inline doc/uftrace-tui.md]: Please reorder as follows.
```suggestion
 * `s`:                   Show uftrace report sorted by KEYs
```
In addition, you can also add same description in `doc/ko/uftrace-tui.md` man page in Korean.

Comment (hjongkim) [reply doc/uftrace-tui.md]: I fixed it as you told me. Please check.

File: cmds/tui.c (hunk 3 above)
Comment (namhyung) [reply cmds/tui.c]: Hmm.. it's not changing the view.  I'd rather say "Sort by the next column (in uftrace report)".

File: cmds/tui.c [hunk 7]
```diff
@@ -142,6 +142,7 @@ static const char *help[] = {
 	"g             Show call graph for this function",
 	"R             Show uftrace report",
 	"r             Show uftrace report for this function",
+	"s             Show uftrace report sorted by KEYs",
```
Comment (honggyukim) [inline cmds/tui.c]: Could you change this like below as @namhyung mentioned?
```suggestion
	"s             Sort by the next column in report",
```

File: doc/uftrace-tui.md [hunk 8]
```diff
@@ -147,6 +147,7 @@ Following keys can be used in the TUI window:
  * `g`:                   Show backtrace and call graph of the current function
  * `R`:                   Show uftrace report
  * `r`:                   Show uftrace report of the current function
+ * `s`:                   Show uftrace report sorted by KEYs
```
Comment (honggyukim) [inline doc/uftrace-tui.md]: Please change it like below.
```suggestion
 * `s`:                   Sort by the next column (in report mode)
```

File: doc/ko/uftrace-tui.md [hunk 9]
```diff
@@ -147,6 +147,7 @@ TUI 창에서 다음과 같은 키들을 사용할 수 있음:
  * `g`:                   현재 함수의 백트레이스와 부분 호출 그래프 창으로 전환
  * `R`:                   uftrace report 창으로 전환
  * `r`:                   현재 함수를 기준으로 uftrace report 창으로 전환
+ * `s`:                   KEY 기준으로 정렬된 uftrace report 창으로 전환
```
Comment (honggyukim) [inline doc/ko/uftrace-tui.md]: Please change it like below.
```suggestion
 * `s`:                   다음 열의 항목으로 정렬 (report 모드에서)
```

**Comment by honggyukim** [comment]
@hjongkim It'd be always better to rebase the commit onto the current master. Please do it when you force push it again.

**Comment by hjongkim** [comment]
I applied squash and rebase to commits. Please confirm.

**Comment by honggyukim** [comment]
@namhyung Are you okay merging this into master?

**Comment by namhyung** [comment]
Yep, looks good!

**Pull request merged**